### DISCUSSION TITLE
Non-blazor WASM build?

### DISCUSSION THREAD

**Discussion by SerratedSharp** [dotnet/runtime#99243] [Q&A] [answered]
I currently use Uno.Wasm.Bootstrap to build a .NET WASM package that can be hosted in the browser and use JS interop features like JSImport/JSObject from the System.Runtime.InteropServices.JavaScript namespace.  This is for client side logic that can be used with traditional sites, regardless of what server side UI platform/hosting/server-side is used.

I'm wondering if there are plans for .NET to natively support a non-Blazor project type for building WASM packages and support for System.Runtime.InteropServices.JavaScript?  Something like the wasi experimental workload Console app template, but that is hostable in JS.

**Comment by maraf** [comment, marked as answer]
Do mean something like `wasmbrowser` template from `wasm-experimental` workload?
https://learn.microsoft.com/en-us/aspnet/core/client-side/dotnet-interop?view=aspnetcore-8.0

**Reply by SerratedSharp** [reply]
Indeed, thank you, I'll take a look at that.

**Reply by SerratedSharp** [reply]
@maraf  Would yall be interested in a PR to expand this article or add a related article with some concrete examples of different interop scenarios? I.e. passing JSObject's across interop layer, calling instance methods on JSObject, returning promises via JSImport'd async/task methods, passing C# Action's to JS to support events, etc.  I've documented alot of these examples in my own recipes repo: https://github.com/SerratedSharp/CSharpWasmRecipes/blob/main/README.md

Some of my approaches are opinionated, but I can distill out the less opinionated ones and I think would be handy for others to see concrete examples. 

Also, I'll point out other articles like this one are titled "with ASP.NET Core Blazor" which can be misleading to those new to .NET WASM, since the System.Runtime.InteropServices.JavaScript namespace is supported under other WASM platforms and isn't specific to Blazor:  https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/import-export-interop?view=aspnetcore-8.0

I'm happy to see y'all are not focusing exclusively on Blazor, as I think being able to implement client side logic in other platforms like what I'm doing with ASP.NET MVC will enable alot of capabilities.

**Reply by maraf** [reply]
cc @pavelsavara

**Reply by maraf** [reply]
> Also, I'll point out other articles like this one are titled "with ASP.NET Core Blazor" which can be misleading to those new to .NET WASM, since the System.Runtime.InteropServices.JavaScript namespace is supported under other WASM platforms and isn't specific to Blazor: https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/import-export-interop?view=aspnetcore-8.0

A similar article is beside blazor in the navigation tree under "Client-side dev" https://learn.microsoft.com/en-us/aspnet/core/client-side/dotnet-interop?view=aspnetcore-8.0.

**Comment by pavelsavara** [comment]
> Would yall be interested in a PR to expand this article or add a related article with some concrete examples of different interop scenarios ?

I think it would be good start to enumerate the scenarios/gaps first. Or you could write article on your blog and we could steal the ideas from it?

cc @guardrex for docs guidance

**Reply by guardrex** [reply]
@SerratedSharp, you can open an issue from the bottom of the article ...

https://learn.microsoft.com/en-us/aspnet/core/client-side/dotnet-interop?view=aspnetcore-8.0

@maraf is correct on the location. That's the sister JS Import/Export article for non-Blazor scenarios, and that's either the general area to place a new article or the exact article where new content would be added.

In your issue, @SerratedSharp, you can detail what you're proposing to add. If the proposal is for a new article, then provide an outline in your comment. I'll ask Dan to look (along with @pavelsavara and @maraf advising us on your proposal), as Dan needs to sign off on new articles. If it's just for new content added to the existing article, then I defer to @pavelsavara and @maraf to call the shots on what's added.
